### PR TITLE
Stable Marriage problem and solution

### PR DESCRIPTION
**Pull request by brogowski**
Hello,

I've implemented solution to Stable Marriage problem based on Gale–Shapley algorithm mentioned in #111.

- [x] I have performed a self-review of my code
- [x] My code follows the style guidelines of this project
- [x] I have added comments to hard-to-understand areas of my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have made corresponding changes to the comments
- [x] I have made corresponding changes to the README.md

### CODE REVIEW

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- Algorithms/Problems/StableMarriage/GaleShapley.cs  8
- Algorithms/Problems/StableMarriage/Accepter.cs  2
- Algorithms.Tests/Problems/StableMarriage/GaleShapleyTests.cs  2
- Algorithms/Problems/StableMarriage/Proposer.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6318000&bid=20538743)

File: Algorithms/Problems/StableMarriage/GaleShapley.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                    var accepter = session.Accepters.First();
+                    var newProposer = session.Proposer;
+
+                    if (accepter.EngagedTo == null)
+                    {
+                        Engage(newProposer, accepter);
+                    }
+                    else
+                    {
+                        if (accepter.PrefersOverCurrent(newProposer))
```
Comment (siriak) [inline Algorithms/Problems/StableMarriage/GaleShapley.cs:48]: Codacy found an issue: [Refactor this code to not nest more than 3 control flow statements.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6318000)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Algorithms.Tests/Problems/StableMarriage/GaleShapleyTests.cs  3
- Algorithms/Problems/StableMarriage/Proposer.cs  1
- Algorithms/Problems/StableMarriage/GaleShapley.cs  5
- Algorithms/Problems/StableMarriage/Accepter.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6318000&bid=20538743)

File: Algorithms/Problems/StableMarriage/GaleShapley.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            {
+                proposer.EngagedTo = null;
+            }
+
+            foreach (var accepter in accepters)
+            {
+                accepter.EngagedTo = null;
+            }
+
+            var matchingSession = proposers.Select(p => (Proposer: p, Accepters: accepters.OrderBy(a => p.Score(a)).ToList())).ToArray();
```
Comment (siriak) [inline Algorithms/Problems/StableMarriage/GaleShapley.cs:34]: There is this list already in   `public List<Accepter> PreferenceOrder { get; set; } = new List<Accepter>();`.
So just need to pass p.

File: Algorithms/Problems/StableMarriage/GaleShapley.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                else
+                {
+                    if (accepter.PrefersOverCurrent(newProposer))
+                    {
+                        Free(accepter.EngagedTo);
+                        Engage(newProposer, accepter);
+                    }
+                }
+
+                session.Accepters.Remove(accepter);
```
Comment (siriak) [inline Algorithms/Problems/StableMarriage/GaleShapley.cs:64]: Accepters is a list, which means that his operation takes O(n) time, which brings total algorithm time to O(n^3), which is not correct. Use LinkedList<Accepter> here.

File: Algorithms/Problems/StableMarriage/GaleShapley.cs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            foreach (var proposer in proposers)
+            {
+                proposer.EngagedTo = null;
+            }
+
+            foreach (var accepter in accepters)
+            {
+                accepter.EngagedTo = null;
+            }
+
```
Comment (siriak) [inline Algorithms/Problems/StableMarriage/GaleShapley.cs:33]: No need for this as they should be null before calling the algorithm

**Review by siriak — CHANGES_REQUESTED** [review]
Overall looks good, I've left a couple of comments on implementation details that are important and make the algorithm run in O(n^3) instead of advertised O(n^2).

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Algorithms/Problems/StableMarriage/Proposer.cs  1
- Algorithms.Tests/Problems/StableMarriage/GaleShapleyTests.cs  3
- Algorithms/Problems/StableMarriage/GaleShapley.cs  4
- Algorithms/Problems/StableMarriage/Accepter.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6318000&bid=20538743)

**Comment by brogowski** [comment]
> Overall looks good, I've left a couple of comments on implementation details that are important and make the algorithm run in O(n^3) instead of advertised O(n^2).

Thanks for catching those 👍 - all implemented already.

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Algorithms/Problems/StableMarriage/GaleShapley.cs  4
- Algorithms.Tests/Problems/StableMarriage/GaleShapleyTests.cs  3
- Algorithms/Problems/StableMarriage/Accepter.cs  2
- Algorithms/Problems/StableMarriage/Proposer.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6318000&bid=20538743)

**Review by siriak — APPROVED** [review]
Thanks!

**Pull request merged**